### DISCUSSION TITLE
General AC element simulation

### DISCUSSION THREAD

**Discussion by ZeusMarti** [atcollab/at#503] [General]
Hi all,

We are considering operating in full coupling with an AC skew magnet. Is there a way to integrate such type of elements in AT and calculate closed orbit, (average) optics functions, emitances and dynamical apertures? 

One could track one turn at a time and change the magnet strength every turn, but I wonder if AT has or is planned to have a built in way to deal with that. If not, any ideas how to proceed? Or do you know it that is implemented in other codes? Elegant maybe? 

Best,

Zeus

**Comment by swhite2401** [comment]
Hi @ZeusMarti,

In fact there are ways to integrate time dependent tracking elements as the turn number is now accessible in the tracking engine.
You can get closed orbit, emittance and DA from tracking but optics functions will be more tricky, most likely you will need to vary the element manually and compute optics in a loop.

@carmignani has an AC dipole module, I think he can share it as an example.
We may nevertheless think of a more general element that could have one or more time-dependent attributes.
Would this be of some use for you? Do you have a deadline for this development?

I don't know if elegant has such feature....

**Comment by ZeusMarti** [comment]
Hi @swhite2401 and everyone,

That is interesting, I'm not familiar with that feature, can we add a field to the element structure containing the kick at every turn? I guess that would be much faster than tracking for one turn in a loop. No deadlines yet, we are in the early stages of the ALBA-II design, beside this is just one of the possibilities, we are not discarding DC full coupling or even vertical AC dipole excitations. Still I think it is quite important to quantify the foreseeable performance of each different approach. So yes, it would be very helpful for me if @carmignani can share his example.

More generally speaking, I'm not sure I understand all the implications of such AC skew element. For example, the closed orbit may exist but somehow we have to get rid of the feeddown AC dipole component in the orbit. That is a key issue since this is the base of the LOCO measurement. Maybe it works if we measure the closed orbit with AC orbit correctors excitation such that the AC skew feed-down effect is filtered out via fft. Or maybe we need to accept that the LOCO measurement has no sense with such element active? Any idea?

I just checked, in Elegant there is this MBUMPER element, which can be any multipole and you can specify a time dependent variation which in particular can be defined as periodic. I just asked in their forum since I'm not sure you can use it to calculate the optics. Any way, since several pieces of the control system use MML and AT it would be much easier if this is development exists in AT.

Thanks a lot,

**Comment by swhite2401** [comment]
Interesting problem, and indeed difficult to imagine, I supposed you'll have to try.

Ok I can take a look, maybe sometimes next week, something similar to MBUMPER should be relatively straight forward. A*sin(w*t) type of excitation is the simplest to implement if that is what you need.

**Comment by ZeusMarti** [comment]
That would be perfect , thanks Simon!

**Comment by swhite2401** [comment]
Hi @ZeusMarti,

So finally I found time to come up with a test version this afternoon. For now it is just a passmethod for you to do some testing.
The code is found on the branch `ac_mpole`, to install it:

```
git checkout ac_mpole
atmexall (for matlab)
or
pip install . (for python)
```


As agreed it is a variable thin lense multipole, the variations are as follows:

```
AmplitudeA*sin(2*pi*Frequency*t)
AmplitudeB*sin(2*pi*Frequency*t)
```

So in principle you can excite any kind of multipole with this. Below an example usage for AC dipole.
Let me know if that is what you were looking for, in which I can produce user interface for element creation and arbitrary
excitation patterns.

Have fun!


Matlab:
```
clear;

v = load('./S28F.mat');
ring = v.LOW_EMIT_RING;
ring = atradoff(ring,'CAVIPASS','RFCavityPass');
frev = atGetRingProperties(ring, 'revolution_frequency');
tunes = tunechrom(ring);

morder = 3;
acm = atbaselem('FamName', 'ACMPOLE', 'PassMethod', 'VariableThinMPolePass');
acm.MaxOrder = morder;
acm.PolynomB = zeros(1,morder);
acm.PolynomA = zeros(1,morder);
acm.AmplitudeB = zeros(1,morder);
acm.AmplitudeA = zeros(1,morder);
acm.AmplitudeB(1)=1e-5;
acm.Frequency = (tunes(1)-0.01)*frev;

ring = [ring;{acm}];

rout = squeeze(ringpass(ring,zeros(6,1),1000));
plot(rout(1,:))
```

Python (using fastring in this case):
```
import numpy
import matplotlib.pyplot as plt
import at


path = '../lattice/'
filename = 'S28F.mat'
key = 'LOW_EMIT_RING_INJ'
latticef = path+filename
ring = at.load_lattice(latticef,key=key)
ring.radiation_off(cavity_pass='CavityPass')
tunes = ring.get_tune()
frev = ring.get_revolution_frequency()

print(tunes)
print(frev)

fring, _ = at.fast_ring(ring)

maxorder = 3
pola = numpy.zeros(maxorder)
polb = numpy.zeros(maxorder)
ampa = numpy.zeros(maxorder)
ampb = numpy.zeros(maxorder)
ampb[0] = 10e-6;
frequency = (0.01)*frev
acmpole = at.Element('ACMPOLE',
                     PassMethod='VariableThinMPolePass',
                     PolynomA=pola, PolynomB=polb,
                     AmplitudeA=ampa, AmplitudeB=ampb,
                     Frequency = frequency, MaxOrder = maxorder)
                     
print(acmpole)
fring.append(acmpole)

xout = numpy.squeeze(at.tracking.lattice_pass(fring,numpy.zeros(6),nturns=1000))
print(acmpole)
plt.plot(xout[0,:])
plt.show()
```

**Comment by carmignani** [comment]
Hello,

I have a thin ACDipole element that I can upload. It is relatively easy to transform it into a thick ACDipole element and it can be used to get some ideas for a AC multiple.

**Comment by carmignani** [comment]
I upload the file here, then if you think it is useful I can maybe extend to think elements and then I can do an official pull request.

It has to be renamed to ACDipole.c and then compiled with atmexall. I did not try it on python, but it should work.
I didn't use it much, so there could be some bugs.

[ACDipolePass.txt](https://github.com/atcollab/at/files/9821546/ACDipolePass.txt)

**Comment by ZeusMarti** [comment]
Wow, that was fast, thanks a lot Simon an Nicola,

I'm testing it, no problems with traking (the action exchange works as expected). I guess the effect on the orbit and the optics is simply ignored (only the first turns are used in that case right?). Maybe a way to generalize those calcullations would be using N turns of the ring where N=frev/fac in the orbit and optics calculation, does it make sense?
Thanks again,
Zeus

**Comment by swhite2401** [comment]
Yes that is correct, only the first turn is used for optics and orbit.
I think what you want is an average over Nturns, for the orbit it is easy just do:

`rout = ringpass(ring,rin,nturns)`

and then average over nturns. I think this is the closest you can get to an integrated orbit measurement.

For optics this is much more diffcult because it is based on transfer maps. You may however compute an orbit response matrix using the average orbit above, fit it a derive the optics from there. This would be representative of what you would actually measure in real life.

Other option is to run consecutive optics calculations with different skew settings and average.

Now from a theoretical standpoint what you propose is interesting but also a completely new definition, as in you are looking at solutions that are periodic over several turns instead of the usual one turn...I am not sure what are the implications of this since our magnets settings are static in any case

**Comment by ZeusMarti** [comment]
Hi Simon and all,

The new pass method works nicely, is it ok to merge it?
Best,

**Comment by swhite2401** [comment]
Hi Zeus,
I fact I was in the process to improve it to integrate arbitrary wave forms, can it wait a little bit?

**Comment by ZeusMarti** [comment]
Ah sure, no problem!

By the way, which would be the easyest way calculate the emittances with such device? Maybe just by tracking it together with a simplified version of the ring (with radiation and diffusion)? Is this possible?

**Comment by carmignani** [comment]
For the AC dipole simulations, we found out that the ramp up and ramp down of the excitation was useful in some cases. I did not check @swhite2401 pass method, but maybe it could be implemented in the AC multiple. 
If the excitation does not start at turn 0, you have also the advantage that the optics computations are the one with no oscillating elements.

**Comment by swhite2401** [comment]
Yes it should be possible to run a "reduced" scenario using `atfastring` that creates a 1 turn map and `atQuantDiff` that creates the diffusion element.
You may also combine `atQuantDiff` with the full ring, this is still reasonable computing time but slower. 

@carmignani , yes I was planning to add it as well. I will make a PR for review once I have something clean. I may need some help with the matlab element creation.... I am really not comfortable with this language

**Comment by swhite2401** [comment]
In fact `atfastring` includes the diffusion element.
So you just have to run:

`[FASTRING,FASTRINGRAD]=ATFASTRING(RING)`

The second output contains everything you need to get the emittance from tracking. It won't be exact thought as this is an approximation, but will give you quick estimates on the effect of your AC skew

**Comment by ZeusMarti** [comment]
Thanks a lot Simon! I'll give it a try it asap.

**Comment by ZeusMarti** [comment]
Hi again Simon,

I just wanted to let you know that I discussed with Laura and Ubaldo yesterday and they are very interested in that arbitrary wave forms you mentioned since they want to study a effect of the white noise for an stable beam blow up. Will that be based on a kind of TbT kick list? Also if we add a field called "periodic" like Elegant's MBUMPER the list could be repeated or not, is that your idea?

In any case, let me know if I can help,

Cheers,
Zeus

**Comment by swhite2401** [comment]
Hi,
I did progress on the topic and I have implemented 3 modes:

- the sine wave that you have tested, with the addition of a phase input and independent frequencies for normal and skew components
- an ideal white noise using the a C gaussian random number generator
- an arbitrary function, that is in reality a vector f(turn number). This table can be of any length, in case it is shorter than the number of tracked turns it is repeated. 

For each of these methods you can optionally specify four numbers to define the start of the linear ramp up, the start of the flat-top the start of the linear ramp-down and the end of the ramp-down

What do you think? Is this matching your needs?

I will soon have something clean enough, in fact the C part is done, to be proposed in a pull request. I will add you as reviewer so you will get a chance to propose modifications and improvements
